python/perforce.py

```diff
@@ -356,30 +356,46 @@ def p4print_unshelve(self, changelist):
         if 'depotFile' not in changeinfo:
             raise Exception('Changelist %s does not contain any shelved files' % changelist)
         depotfiles = changeinfo['depotFile']
+
+        # Filter out files with exclusive locking
+        filtered_files = []
+        exclusive_files = []
+        for i in range(len(depotfiles)):
+            if "+l" in changeinfo["type"][i]:
+                exclusive_files.append(depotfiles[i])
+                continue
+            filtered_files.append(depotfiles[i])
         
-        whereinfo = self.perforce.run_where(depotfiles)
+        cmds = []
 
-        depot_to_local = {item['depotFile']: item['path'] for item in whereinfo}
+        if filtered_files:
+            # Run unshelve on non-exclusive files. Exclusive locking files
+            # are p4 printed instead.
+            cmds.append(('unshelve', '-s', changelist, filtered_files))
```

Comment:
We need to run a second filter on this to only sync files in specified paths, similar to what we do below with the `print`s.

Reply:
Not sure I'm following. I'm providing the list of files in the shelf with `filtered_files`

Reply:
A given buildkite step can have partial sync paths (for efficiency on a given build step that doesn't need the whole depot), e.g.
- //depot/stream/very/specific/context/...
- //depot/stream/other/path/with/dependencies/...

In the p4 print process, we filter for these prefixes and only `print` things that belong to these trees; We should do the same for this unshelve operation, so we don't try to unshelve paths within the shelf that the build machine doesn't have (because it's not in the sync paths).

Reply:
Ah I see. We should provide a list of common prefixes:
```
            sync_prefixes = [prefix.rstrip('.') for prefix in self.sync_paths]

            synced_patched_files = []

            for depotfile, localfile in depot_to_local.items():
                if any(depotfile.startswith(prefix) for prefix in sync_prefixes):
                    if os.path.isfile(localfile):
                        os.chmod(localfile, stat.S_IWRITE)
                        os.unlink(localfile)
```

Reply:
Updated in latest

Reply:
Test preflight https://buildkite.com/inflexion/nwx-dev-ci/builds/128770#0192b040-bb71-47b2-9bd7-2c60508764c0